File: algorithms/0079.word_search/word_search.rs
type Table<T> = Vec<Vec<T>>;
impl Solution {
    
    pub fn exist(board: Table<char>, word: String) -> bool {
        let mut visited = vec![vec![false; board[0].len()]; board.len()];
        let word = word.chars().collect::<Vec<char>>();
        for i in 0..board.len() {
            for j in 0..board[0].len() {
                if Self::backtracking(&board, &word, &mut visited, 0, i, j) {
                    return true;
                }
            }
        }
        false
    }

    fn backtracking(
        board: &Table<char>,
        word: &Vec<char>,
        visited: &mut Table<bool>,
        k: usize,
        i: usize,
        j: usize,
    ) -> bool {
        if i >= board.len() || j >= board[0].len() || visited[i][j] {
            return false;
        }
        if board[i][j] != word[k] {
            return false;
        }
        visited[i][j] = true;
        if k == word.len() - 1 {
            return true;
        }
        // println!("{:?} {} k:{}", (i, j), board[i][j], k);
        if (j != 0 && Self::backtracking(board, word, visited, k + 1, i, j - 1))
            || Self::backtracking(board, word, visited, k + 1, i + 1, j)
            || (i != 0 && Self::backtracking(board, word, visited, k + 1, i - 1, j))
            || Self::backtracking(board, word, visited, k + 1, i, j + 1)
        {
            return true;
        }
        visited[i][j] = false;
        false
    }

}

File: algorithms/0079.word_search/word_search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string& word) {
        const int m = board.size(), n = board[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == word[0] && dfs(board, i, j, word, 0, visited)) {
                    return true;
                }
            }
        }
        return false;
    }
    bool dfs(vector<vector<char>>& board, int i, int j, string& word, size_t k, vector<vector<bool>>& visited) {
        if (k >= word.size()) {
            return true;
        }
        const int m = board.size(), n = board[0].size();
        if (i < 0 || i >= m || j < 0 || j >= n || visited[i][j] || board[i][j] != word[k]) {
            return false;
        }
        visited[i][j] = true;
        bool ret = dfs(board, i + 1, j, word, k + 1, visited)
                    || dfs(board, i - 1, j, word, k + 1, visited)
                    || dfs(board, i, j + 1, word, k + 1, visited)
                    || dfs(board, i, j - 1, word, k + 1, visited);
        visited[i][j] = false;
        return ret;
    }
};